subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c

```diff
@@ -0,0 +1,617 @@
+	/* Trim length based on CONFIG_MCUMGR_GRP_IMG_UPLOAD_DATA_ALIGNMENT_SIZE */
+	if (max_data_length % CONFIG_MCUMGR_GRP_IMG_UPLOAD_DATA_ALIGNMENT_SIZE) {
+		max_data_length -=
+			(max_data_length % CONFIG_MCUMGR_GRP_IMG_UPLOAD_DATA_ALIGNMENT_SIZE);
+	}
+
+	while (request_length != wrote_length) {
+		write_ptr = data + wrote_length;
+		write_length = request_length - wrote_length;
+		if (write_length > max_data_length) {
+			write_length = max_data_length;
+		}
+
+		nb = smp_client_buf_allocation(active_client->smp_client, MGMT_GROUP_ID_IMAGE,
+					       IMG_MGMT_ID_UPLOAD, MGMT_OP_WRITE);
+		if (!nb) {
+			image_upload_buf->status = MGMT_ERR_ENOMEM;
+			goto end;
+		}
+
+		zcbor_new_encode_state(zse, ARRAY_SIZE(zse), nb->data + nb->len,
+				       net_buf_tailroom(nb), 0);
+		if (active_client->upload.offset) {
+			map_count = 6;
+		} else if (active_client->upload.hash_initialized) {
+			map_count = 12;
+		} else {
+			map_count = 10;
+		}
+
+		/* Init map start and write image info and data */
+		ok = zcbor_map_start_encode(zse, map_count) && zcbor_tstr_put_lit(zse, "image") &&
+		     zcbor_uint32_put(zse, active_client->upload.image_num) &&
+		     zcbor_tstr_put_lit(zse, "data") &&
+		     zcbor_bstr_encode_ptr(zse, write_ptr, write_length);
+		/* Write Len when offset is zero */
+		if (ok && !active_client->upload.offset) {
+			ok = zcbor_tstr_put_lit(zse, "len") &&
+			     zcbor_size_put(zse, active_client->upload.image_size);
+		}
+		/* Write offset */
+		if (ok) {
+			ok = zcbor_tstr_put_lit(zse, "off") &&
+			     zcbor_size_put(zse, active_client->upload.offset);
+		}
+		/* Write hash when it defined and offset is 0 */
+		if (ok && !active_client->upload.offset && active_client->upload.hash_initialized) {
+			ok = zcbor_tstr_put_lit(zse, "sha") &&
+			     zcbor_bstr_encode_ptr(zse, active_client->upload.sha256,
+						   IMG_MGMT_HASH_LEN);
+		}
+
+		if (ok) {
+			ok = zcbor_map_end_encode(zse, map_count);
+		}
+
+		if (!ok) {
+			LOG_ERR("Failed to encode Image Upload packet");
+			smp_packet_free(nb);
+			image_upload_buf->status = MGMT_ERR_ENOMEM;
+			goto end;
+		}
+
+		offset_before_send = active_client->upload.offset;
+		nb->len = zse->payload - nb->data;
+		k_sem_reset(&spm_mutex);
+
+		image_upload_buf->status = MGMT_ERR_EINVAL;
+		image_upload_buf->image_upload_offset = SIZE_MAX;
+
+		rc = smp_client_send_cmd(active_client->smp_client, nb, image_upload_res_fn,
+					 &spm_mutex, MCUMGR_MGMT_FLASH_OPERATION_TIMEOUT);
+		if (rc) {
+			LOG_ERR("Failed to send SMP Upload init packet, err: %d", rc);
+			smp_packet_free(nb);
+			image_upload_buf->status = rc;
+			goto end;
+
+		}
+		k_sem_take(&spm_mutex, K_FOREVER);
+		if (image_upload_buf->status) {
+			LOG_ERR("Upload Fail:%d", image_upload_buf->status);
+			goto end;
+		}
+
+		if (offset_before_send + write_length < active_client->upload.offset) {
+			/* Offset further than expected which indicate upload session resume */
+			goto end;
```

Comment:
Network client might get an 8KB chunk, might save it here, it might get interrupted, it reconnects and gets a 32KB chunk, writes the first packet and it hits this part and exits until it rights the next part... What happens to the extra 32-8 = 24KB of data from this packet which was not written?

Reply:
OK. In that case we need to download same data again based on server reported offset. Coap use 512 bytes and http 1024 for requesting data.

Reply:
Is this always the case or can it change, i.e. if you are one cellular network and you disconnect then connect to another, could the size of the data you receive or get this callback on be different (or does the web client buffer it)?

Reply:
Download client config received data size to request.